File: 14a/main.py
from collections import Counter


with open('input.txt', 'rt') as f:
    lines = [l.strip() for l in f.readlines()]
template = lines[0]
rules = {
    tuple(l.split(' -> ')[0]): l.split(' -> ')[1]
    for l in lines[2:]
}

for _ in range(10):
    pairs = zip(template, template[1:])
    insertions = []
    for index, pair in enumerate(pairs):
        if pair in rules:
            insertions.append(rules[pair])
        else:
            insertions.append('')
    # Insertions will have one fewer element.
    insertions.append('')
    new_str = ''.join(i + j for i, j in zip(template, insertions))
    template = new_str

counter = Counter(template)
freqs = counter.values()
result = max(freqs) - min(freqs)

print(result)

File: 14a/ruby/main.rb
FILENAME = "input.txt"

lines = File.open(FILENAME).readlines.map(&:chomp)

sequence = lines[0]
rules = Hash[
  lines[2..].map do |line|
    pair, insertion = line.split(" -> ")
  end
]

class String
  def update(rules)
    new_seq = split("").join(" ")
    (length-1).times do |i|
      new_seq[2*i+1] = rules[self[i]+self[i+1]]
    end
    new_seq
  end
end

10.times do
  sequence = sequence.update(rules)
end
seq_array = sequence.split("")
most_common_letter = seq_array.max_by{|s| seq_array.count(s)}
least_common_letter = seq_array.min_by{|s| seq_array.count(s)}
result = seq_array.count(most_common_letter) - seq_array.count(least_common_letter)
p result
